### ISSUE TITLE
[FEA] round-robin partitioning

### ISSUE THREAD

**Issue by jlowe** [rapidsai/cudf#3648]
**Is your feature request related to a problem? Please describe.**
I would like to evenly distribute data between partitions.  Spark has a round-robin partitioner designed to do this, and it would be great if this could be GPU-accelerated.

**Describe the solution you'd like**
A table API that is similar to the existing hash partition that partitions in a round-robin fashion based on row index rather than hashing the row values.  I would also like to be able to specify the starting partition number rather than always assigning the first row of the table to partition 0.  Something like:
```c++
/**
 * Returns a new table with rows re-arranged into partition groups and
 * a vector of row offsets to the start of each partition in the output table.
 * Rows are assigned partitions based on their row index in the table,
 * based on the following formula:
 * partition = (row_index + start_partition) % num_partitions
 */
std::pair<std::unique_ptr<table>, std::vector<size_type>>
round_robin_partition(table_view const& input,
                      size_type num_partitions,
                      size_type start_partition = 0, 
                      rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
                      cudaStream_t stream = 0);
```

**Describe alternatives you've considered**
Simply slicing the table into N partitions evenly distributes the rows across partitions, but this often will not evenly distribute the values.  Spark always ensures the table is sorted before round-robin partitioning to achieve idempotent behavior for fault tolerance.  Slicing a sorted table will often concentrate specific key values to a small number of tasks leading to large data skew in downstream processing.

Round-robin partitioning could also be implemented with `scatter_to_tables` although there are a couple of drawbacks:
- the API requires manifesting the mapping column when this could be avoided with a slightly modified counting iterator for the map.
- `scatter_to_tables` allocates separate tables which is undesired as detailed in #3566

**Comment by aschaffer** [comment]
Some important specs / examples:

"As for the semantics for the start_partition argument, it controls which partition we start populating with row values in the round-robin algorithm. Using the same example above:

Take nrows = 11, npartitions = 3
If you round-robin [0,1,2,3,4,5,6,7,8,9,10] with start_partition=0 then the resulting table arrangement needs to be:
[0,3,6,9,1,4,7,10,2,5,8]
and the partition offsets vector is:
[0,4,8]

If the start_partition changes to 1 then whatever used to be in partition 0 is now in partition 1, partition 1 moves to 2, etc. It is like a rotate right, but the rotation occurs in the table arrangement rather than the offsets vector. The output needs to have a table arrangement of:
[2,5,8,0,3,6,9,1,4,7,10]
and the partition offsets vector is:
[0,3,7]"

**Comment by aschaffer** [comment]
Additional requirements (quoting @jlowe, 1/15/2020):

It would be nice if the partitioner handled num_partitions >= nrows. In the case where num_partitions == nrows, each partition has one row. In the case where num_partitions > nrows, some partitions contain zero rows. I realize these are sort of degenerate cases, but let me explain the use-case where it can make sense from the caller's perspective. Apologies for the wall of text.

Partitioning is always followed by a split call in preparation for transferring the partitions to different downstream tasks. Round robin partitioning is just another partitioning algorithm like hash partitioning. Arguably round-robin partitioning is just another form of hash partitioning where the hash function is the row number plus a specified offset. Hash partitioning supports num_partitions >= nrows, which is why I'd like to see it supported here. That way the caller can always partition followed by a call to split and it works, regardless of the partitioning algorithm used. Otherwise the caller needs to check for the round-robin num_partitions >= nrows case and handle splitting differently to avoid the code exploding.

Regarding how this would work in practice, here's an example for num_partitions == nrows:
With input table => col 1 {0, ..., 10}, num_partitions = 11, and start_partition = 2, the result should be:
table => col 1 {9, 10, 0, 1, 2, 3, 4, 5, 6, 7, 8}
offsets => {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
The table needs to be transformed when start_partition > 0, but the offsets are just a counting vector.

For the num_partitions > nrows case, take the same example with num_partitions=15. We start handing out row 0 to partition 2, row 1 to partition 3, etc. Eventually we run out of rows after partition 12, so partitions 13, 14, 0, and 1 would be empty. This would be represented by the following output:
table => col 1 { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 }
offsets => {0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 11 }

As another num_partitions > nrows case, consider:
input table => col 1 {0, ..., 10}, num_partitions = 15, and start_partition = 10.
We start handing out row 0 to partition 10, row 1 to partition 11, etc. Eventually we wrap partitions and run out of rows after partition 5, so partitions 6, 7, 8, and 9 would be empty. This would be represented by the following output:
table => col 1 {5, 6, 7, 8, 9, 10, 0, 1, 2, 3, 4}
offsets => {0, 1, 2, 3, 4, 5, 6, 6, 6, 6, 6, 7, 8, 9, 10}

These transforms could be handled by the caller, possibly more efficiently if the offsets don't need to be manifested and the table copied. However this gets complicated for the caller when the splitting is via `contiguous_split`, as we want all the column data in a partition grouped together for more efficient transfers. `contiguous_split` cannot be called without a properly table ordering and offsets vector. Therefore it significantly simplifies the caller logic if the data can always be partitioned and split without needing to check the specific partition algorithm used and boundary cases of that algorithm.



>>    start_partition also can be larger than num_rows. I'm not sure how that makes sense.

start_partition specifies the starting partition, not row, of the round-robin algorithm.
For example, input table => col 1 {0, ..., 10}, num_partitions = 15, and start_partition = 14 would result in:
table => col 1 {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 0}
offsets => {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10, 10}

This indicates partitions 10, 11, 12, and 13 are empty since start offset == end offset for those partitions. Whenever num_partitions >= nrows, the table output will be some rotation of the input table, and the offsets vector will be a counting vector with between 0 and 2 strings of duplicate entries to mark the ranges of empty partitions.



num_partitions > nrows is handled by generating empty partitions.

Think of the algorithm as dealing out cards.

    the deck of cards is represented as the rows in the table.
    the number of partitions is the number of players being dealt cards.
    the start_partition indicates which player starts getting cards first.

The algorithm result is two things:

    Another deck of cards formed by stacking each player's cards back into a deck again, preserving the order of cards dealt to each player, starting with player 0.
    A vector into the output deck indicating where a player's cards start.

A player's deck (partition) is the range of cards starting at the corresponding offset and ending at the next player's starting offset or the last card in the deck if it's the last player.

When num_partitions > nrows, we have more players than decks. We start dealing to the first indicated player and continuing around the players until we run out of cards before we run out of players. Players that did not get any cards are represented by
`offset[i] == offset[i+1]` or
`offset[i] == table.num_rows()` if `i == num_partitions-1`
meaning there are no cards (rows) in their deck (partition).

Hopefully with that analogy the examples I gave earlier for the case of num_partitions > nrows will make sense.

**Comment by aschaffer** [comment]
Links of interest (some of which duplicate the above post):

1. https://github.com/rapidsai/cudf/pull/3667#discussion_r366914838
2. https://github.com/rapidsai/cudf/pull/3667#discussion_r366927777
3. https://github.com/rapidsai/cudf/pull/3667#discussion_r366971874
4. https://github.com/rapidsai/cudf/pull/3667#discussion_r366974019
5. https://github.com/rapidsai/cudf/pull/3667#discussion_r367039964
6. https://github.com/rapidsai/cudf/pull/3667#discussion_r367054472
7. https://github.com/rapidsai/cudf/pull/3667#discussion_r367067229
8. https://github.com/rapidsai/cudf/pull/3667#discussion_r367097226

**Comment by jrhemstad** [comment]
Closed by #3667
